### PR TITLE
wallet, bench: Move commonly used functions to their own file and fix a bug

### PR DESCRIPTION
**Pull request by achow101**
I have a few PRs and branches that use these two commits, probably makes sense to split them into a separate PR to be merged sooner.

The first commit contains some things that end up being commonly used in new wallet benchmarks. These are moved into `wallet_common.{h/cpp}`.

The second commit contains a bugfix for the wallet_balance benchmark where it calls `LoadWallet` in the wrong place. It's unnecessary to call that function in this benchmark. Although this does not cause any issues currently, it ends up causing issues in some PRs and branches that I'm working on.

### CODE REVIEW

File: src/bench/util/wallet_common.cpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+std::shared_ptr<CWallet> BenchLoadWallet(std::unique_ptr<WalletDatabase> database, WalletContext& context, uint64_t create_flags)
+{
+    bilingual_str error;
+    std::vector<bilingual_str> warnings;
+    auto wallet = CWallet::Create(context, "", std::move(database), create_flags, error, warnings);
+    NotifyWalletLoaded(context, wallet);
+    if (context.chain) {
+        wallet->postInitProcess();
+    }
```
Comment (furszy) [inline src/bench/util/wallet_common.cpp:17]: little topic:
it seems odd to re-use this two calls in other bench contexts.

`NotifyWalletLoaded` is looping over the wallet load callbacks, and `postInitProcess` syncs the wallet with the mempool, so guess that both will be no-op in most cases?.

Comment (achow101) [reply src/bench/util/wallet_common.cpp:17]: Yes, I think they'll be no-op for the most part.

File: src/bench/util/wallet_common.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+#ifndef BITCOIN_BENCH_UTIL_WALLET_COMMON_H
+#define BITCOIN_BENCH_UTIL_WALLET_COMMON_H
+
+#include <wallet/context.h>
+#include <wallet/wallet.h>
+
+using wallet::CWallet;
+using wallet::DatabaseOptions;
+using wallet::WalletContext;
+using wallet::WalletDatabase;
```
Comment (furszy) [inline src/bench/util/wallet_common.h:14]: tiny nit: could wrap the functions around a "namespace wallet {}" instead.

**Review by furszy — COMMENTED** [review]
Code ACK, small last topic:

Seems that `BenchLoadWallet` and `BenchUnloadWallet` are similar to `TestLoadWallet` and `TestUnloadWallet` from `wallet_tests.cpp` . The only diff seems to be in the db.

Maybe instead of creating a new file, we could move the bench functions to `wallet/test/util.h` and `wallet/test/util.cpp` so they are used equally for tests and benchmarks?

**Comment by achow101** [comment]
> Maybe instead of creating a new file, we could move the bench functions to `wallet/test/util.h` and `wallet/test/util.cpp` so they are used equally for tests and benchmarks?

Good point, made that change.

It required a few changes to the `CreateWallet` test since that wasn't expecting `postInitProcess` to be run, but should be straightforward to review.

File: src/bench/util/wallet_common.h (hunk 2 above)
Comment (achow101) [reply src/bench/util/wallet_common.h]: Dropped `wallet_common.h`, but still ended up doing this in `wallet_loading` and `wallet_balance`

File: src/wallet/test/wallet_tests.cpp [hunk 3]
```diff
@@ -885,7 +866,9 @@ BOOST_FIXTURE_TEST_CASE(CreateWallet, TestChain100Setup)
             SyncWithValidationInterfaceQueue();
         });
     wallet = TestLoadWallet(context);
-    BOOST_CHECK_EQUAL(addtx_count, 2);
+    // Since mempool transactions are requested at the end of loading, there will
+    // be 2 additional AddToWallet calls, one from the previous test, and a duplicate for mempool_tx
+    BOOST_CHECK_EQUAL(addtx_count, 2 + 2);
```
Comment (furszy) [inline src/wallet/test/wallet_tests.cpp:871]: Not for this PR as it touches the wallet internals but the duplicated `AddToWallet` calls shouldn't be needed if the wallet wtx state is equal to the one in the event.

**Review by furszy — APPROVED** [review]
ACK 7379a54

**Comment by fanquake** [comment]
@furszy looks like you ACK'd the wrong commit here.

**Comment by furszy** [comment]
oops, fixed. That was the one from the comment. Thanks @fanquake.

**Pull request merged**